sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp

```diff
@@ -254,43 +263,51 @@ class annotated_ptr {
 template <typename T, typename... Props>
 class __SYCL_SPECIAL_CLASS
 __SYCL_TYPE(annotated_ptr) annotated_ptr<T, detail::properties_t<Props...>> {
-  using property_list_t = detail::properties_t<Props...>;
 
   static_assert(std::is_same_v<T, void> || std::is_trivially_copyable_v<T>,
                 "annotated_ptr can only encapsulate either "
                 "a trivially-copyable type "
                 "or void!");
 
-  // buffer_location and alignment are allowed for annotated_ref
-  // Cache controls are allowed for annotated_ptr
-  using allowed_properties =
-      std::tuple<decltype(ext::intel::experimental::buffer_location<0>),
-                 decltype(ext::oneapi::experimental::alignment<0>),
-                 decltype(ext::intel::experimental::read_hint<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>),
-                 decltype(ext::intel::experimental::read_assertion<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>),
-                 decltype(ext::intel::experimental::write_hint<
-                          ext::intel::experimental::cache_control<
-                              ext::intel::experimental::cache_mode::cached,
-                              cache_level::L1>>)>;
-  using filtered_properties =
-      typename PropertiesFilter<allowed_properties, Props...>::tuple;
-
-  // template unpack helper
-  template <typename... FilteredProps> struct unpack {};
+  using property_list_t = detail::properties_t<Props...>;
 
-  template <typename... FilteredProps>
-  struct unpack<std::tuple<FilteredProps...>> {
-    using type = detail::properties_t<FilteredProps...>;
+  // annotated_ref type
+  using reference =
+      sycl::ext::oneapi::experimental::annotated_ref<T, property_list_t>;
+
+  // operator enable/disable check
+  enum operator_id {
+    op_plus,
+    op_minus,
+    op_subscript,
+    op_inc,
+    op_dec,
   };
 
-  using reference = sycl::ext::oneapi::experimental::annotated_ref<
-      T, typename unpack<filtered_properties>::type>;
+#define OP_NOT_SUPPORTED(op, property)                                         \
+  "operator" op " is not available when " property " is specified!"
+
+  static constexpr void operatorAvailablityCheck(constexpr unsigned op_id) {
+    static constexpr bool hasAlign =
+        detail::ContainsProperty<alignment_key, std::tuple<Props...>>::value;
+    switch (op_id) {
+    case op_plus:
+      static_assert(!hasAlign, OP_NOT_SUPPORTED("+", "alignment"));
```

Comment:
Should this be a static_assert? In general using static_assert to disable methods is bad. Because it makes it impossible to call as intended but it still appears to be available to SFINAE making certain generic programming impossible.

Usually it's recommended to use enable_if instead. The disadvantage of enable_if is that we can't have a custom error message. But one can generate reasonable good error message with a bit care. E.g. https://godbolt.org/z/bq5qqzh86 produces: "note: candidate template ignored: requirement '!true' was not satisfied [with disabled_by_alignment = true]". Which seems reasonable to me. And as long as we have a reasonable error message, I think, we shouldn't accept the downside of static_assert.

Reply:
I think static_assert generates the most clear error message. 

The solution with "note" could be buried by other overload and not straightforward

Reply:
I agree that the static_assert generates the most clear error message. 

But I think the operator should be disabled by the property. So that SFINAE works and a user-defined overload is chosen (if it is exists). In case I'm not being clear: https://stackoverflow.com/questions/35033306/what-does-it-mean-when-one-says-something-is-sfinae-friendly

Reply:
I see your point. Hmmm, then it means we can never get a hard error here, which will be the best for error message

Reply:
Actually, will this overload, i.e. "operator+" show up in other overload's matching process?

Reply:
> Actually, will this overload, i.e. "operator+" show up in other overload's matching process?
No. Whether it is a member function or a hidden friend (see my other comment) it only participates if at least one operand is an `annotated_ptr`. And that doesn't depend on whether the operator is templated.

BTW: If we were to decide that we use the `static_assert` we would need to update the spec. Currently it says "Operator might be disabled". We use that wording to mean "it is not available". Instead we would need to change that to something like "Calling operator might be ill-formed".

Reply:
Tried the SFINAE approach as follows:
```
template<bool has_alignment=detail::ContainsProperty<alignment_key, std::tuple<Props...>>::value, class=std::enable_if_t<!has_alignment>>
  reference operator[](std::ptrdiff_t idx) const noexcept {
    return reference(m_Ptr + idx);
  }
```
the error message becomes more confusing:
```
test.cpp:28:7: error: attempt to use a deleted function
   28 |       a[i + 2] = a[i + 1] + a[i];
      |       ^
../include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:420:3: note: 'operator int *' has been explicitly marked deleted here
  420 |   operator T *() const noexcept = delete;
      |   ^
```
 this is because when `annotated_ptr::operator[]` is disabled, the compiler tries implicit conversion on the variable `a` to another type whose `operator[]` is available

Reply:
Do we need the `operator T *() const noexcept = delete;`? We have had this since the initial revision of the spec (https://github.com/intel/llvm/blob/b423e1db4077065cf06a160310776f1fcd7c3653/sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc?plain=1#L286). I can't see any other reason than emphasising that conversion isn't supported - but that doesn't seem a good reason. If we remove that the error message of the SFINAE approach should be better.

Also in a small reproducer (https://godbolt.org/z/M3jGeWjGq) I don't see the error you get. But instead get the expected "overload resolution selected deleted operator '[]'". Why do you get the other error about `T*` conversion and not the error about `operator[]` being deleted?

I'm OK with keeping the `static_assert` for now and answering these two questions after it is merged and if either leads to a solution change it in a follow up PR.

Reply:
Hi @rolandschulz, it turns out I missed the keyword `const noexcept` for the deleted `operator[]`, so the signature doesn't match. The SFINAE approach works as we expect. The complete message is as follows:
```
operator_disabled_by_alignment.cpp:26:8: error: overload resolution selected deleted operator '[]'
   26 |       a[i + 2] = a[i + 1] + a[i];
      |       ~^~~~~~
/tmp/wangdi4/frontend/w3/llvm/build/bin/../include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:378:8: note: candidate function [with has_alignment = true, $1 = void] has been explicitly deleted
  378 |   auto operator[](std::ptrdiff_t idx) const noexcept -> decltype("operator[] is not available when alignment is specified!") = delete;
      |        ^
operator_disabled_by_alignment.cpp:26:8: note: built-in candidate operator[](int *, long)
   26 |       a[i + 2] = a[i + 1] + a[i];
      |        ^
operator_disabled_by_alignment.cpp:26:8: note: built-in candidate operator[](const int *, long)
operator_disabled_by_alignment.cpp:26:8: note: built-in candidate operator[](volatile int *, long)
operator_disabled_by_alignment.cpp:26:8: note: built-in candidate operator[](const volatile int *, long)
/tmp/wangdi4/frontend/w3/llvm/build/bin/../include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:372:13: note: candidate template ignored: requirement '!true' was not satisfied [with has_alignment = true]
  372 |   reference operator[](std::ptrdiff_t idx) const noexcept {
      |             ^
```
I've changed the spec back to **"operator is disabled"**.

Reply:
Great!